core/src/main/java/org/apache/iceberg/PartitionSpecUpdate.java

```diff
@@ -0,0 +1,233 @@
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addBucketField(String sourceName, int numBuckets) {
+    newSpecFields.add(builder -> builder.bucket(sourceName, numBuckets));
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addTruncateField(String sourceName, int width, String targetName) {
+    newSpecFields.add(builder -> builder.truncate(sourceName, width, targetName));
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addTruncateField(String sourceName, int width) {
+    newSpecFields.add(builder -> builder.truncate(sourceName, width));
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec renameField(String name, String newName) {
+    Preconditions.checkArgument(curSpecFields.containsKey(name),
+        "Cannot find an existing partition field with the name: %s", name);
+    Preconditions.checkArgument(newName != null && !newName.isEmpty(),
+        "Cannot use empty or null partition name: %s", newName);
+    Preconditions.checkArgument(!curSpecFields.containsKey(newName),
+        "Cannot use partition name more than once: %s", newName);
+
+    PartitionField field = curSpecFields.remove(name);
+    curSpecFields.put(newName,
+        new PartitionField(field.sourceId(), field.fieldId(), newName, field.transform()));
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec removeField(String name) {
+    Preconditions.checkArgument(curSpecFields.containsKey(name),
+        "Cannot find an existing partition field with the name: %s", name);
+    if (base.formatVersion() == 1) {
+      PartitionField field = curSpecFields.remove(name);
+      String newName = field.name() + "_removed"; // rename it for soft delete
+      curSpecFields.put(newName,
+          new PartitionField(field.sourceId(), field.fieldId(), newName, Transforms.alwaysNull()));
+    } else {
+      curSpecFields.remove(name);
+    }
+    return this;
+  }
+
+  @Override
+  public void commit() {
+    TableMetadata update = base.updatePartitionSpec(freshSpecFieldIds(apply()));
+    ops.commit(base, update);
+  }
+
+  private PartitionSpec freshSpecFieldIds(PartitionSpec partitionSpec) {
+    if (base.formatVersion() == 1) {
+      return partitionSpec;
+    }
+
+    int lastAssignedFieldId = 0;
+    Map<String, Integer> partitionFieldIdByName = new HashMap<>();
+    for (PartitionSpec spec : base.specs()) {
+      for (PartitionField field : spec.fields()) {
+        partitionFieldIdByName.put(getKey(field), field.fieldId());
+      }
+      lastAssignedFieldId = Math.max(lastAssignedFieldId, spec.lastAssignedFieldId());
+    }
+
+    PartitionSpec.Builder specBuilder = PartitionSpec.builderFor(base.schema())
+        .withSpecId(partitionSpec.specId());
+
+    for (PartitionField field : partitionSpec.fields()) {
+      int assignedFieldId = partitionFieldIdByName.containsKey(getKey(field)) ?
+          partitionFieldIdByName.get(getKey(field)) : ++lastAssignedFieldId;
```

Comment:
We don't use the return value of a `++` expression to avoid confusion. Incrementing should happen on a separate line, either before or after the value is used.